### PR TITLE
Enable systemd watchdog monitoring for tedge-watchdog daemon

### PR DESCRIPTION
**Pull request by PradeepKiruvale**
Signed-off-by: Pradeep Kumar K J <pradeepkumar.kj@softwareag.com>

## Proposed changes

Enable systemd watchdog monitoring for tedge-watchdog daemon.

- Read the `watchdogsec` from service file, if its present then enable the `watchdog`, and send the `health notification` to the `systemd` on regular intervals.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Review by rina23q — COMMENTED** [review]
The code looks good. Honestly, I am not confident enough to approve as I don't know the watchdog feature so well. I give all my right for approval to Albin.

File: crates/core/tedge_watchdog/src/systemd_watchdog.rs [hunk 1]
```diff
@@ -29,6 +29,37 @@ pub async fn start_watchdog(tedge_config_dir: PathBuf) -> Result<(), anyhow::Err
...
+                tokio::time::sleep(tokio::time::Duration::from_secs(interval / 4)).await;
+            },
+
+            Err(_) => {
+                warn!(
+                    "Watchdog is not enabled for tedge-watchdog: {}",
+                    WatchdogError::NoWatchdogSec {
+                        file: "tedge-watchdog.service".to_string()
+                    }
+                );
```
Comment (didier-wenzek) [inline crates/core/tedge_watchdog/src/systemd_watchdog.rs:56]: Why ignore the error with `_` then to rebuild it as `WatchdogError::NoWatchdogSec { ... }`?

File: crates/core/tedge_watchdog/src/systemd_watchdog.rs [hunk 2]
```diff
@@ -29,6 +29,37 @@ pub async fn start_watchdog(tedge_config_dir: PathBuf) -> Result<(), anyhow::Err
...
+    // Send heart beat notifications to systemd, to notify about its own health status
+    start_watchdog_for_self().await;
+
+    // Monitor health of tedge services
+    start_watchdog_for_tedge_services(tedge_config_dir).await;
+    Ok(())
+}
+
+async fn start_watchdog_for_self() {
+    tokio::spawn(async move {
```
Comment (didier-wenzek) [inline crates/core/tedge_watchdog/src/systemd_watchdog.rs:41]: I would spawn just the loop.

File: crates/core/tedge_watchdog/src/systemd_watchdog.rs [hunk 3]
```diff
@@ -29,6 +29,37 @@ pub async fn start_watchdog(tedge_config_dir: PathBuf) -> Result<(), anyhow::Err
...
+        match get_watchdog_sec("/lib/systemd/system/tedge-watchdog.service") {
+            Ok(interval) => loop {
+                let _ = notify_systemd(process::id(), "WATCHDOG=1").map_err(|e| {
+                    eprintln!("{}", e);
```
Comment (albinsuresh) [inline crates/core/tedge_watchdog/src/systemd_watchdog.rs]:
```suggestion
                    error!("Notifying systemd failed with {}", e);
```

File: crates/core/tedge_watchdog/src/systemd_watchdog.rs (hunk 2 above)
Comment (PradeepKiruvale) [reply crates/core/tedge_watchdog/src/systemd_watchdog.rs]: done!.

File: crates/core/tedge_watchdog/src/systemd_watchdog.rs [hunk 4]
```diff
@@ -29,6 +29,37 @@ pub async fn start_watchdog(tedge_config_dir: PathBuf) -> Result<(), anyhow::Err
...
+                });
+                tokio::time::sleep(tokio::time::Duration::from_secs(interval / 4)).await;
+            },
+
+            Err(_) => {
```
Comment (albinsuresh) [inline crates/core/tedge_watchdog/src/systemd_watchdog.rs]: The error handling should be different for both kinds of errors from `get_watchdog_sec` function. If the error is `ParseWatchdogSecToInt` it's a clear misconfiguration and hence it'd even be okay to fail here without starting the service. If it is `NoWatchdogSec`, then we can continue with a warn.

Comment (PradeepKiruvale) [reply crates/core/tedge_watchdog/src/systemd_watchdog.rs]: done

File: crates/core/tedge_watchdog/src/systemd_watchdog.rs [hunk 5]
```diff
@@ -29,6 +29,42 @@ pub async fn start_watchdog(tedge_config_dir: PathBuf) -> Result<(), anyhow::Err
...
+                        eprintln!("Notifying systemd failed with {}", e);
+                    });
+                    tokio::time::sleep(tokio::time::Duration::from_secs(interval / 4)).await;
+                }
+            });
+            Ok(())
+        }
+
+        Err(e) => match e {
+            WatchdogError::NoWatchdogSec { file } => {
```
Comment (didier-wenzek) [inline crates/core/tedge_watchdog/src/systemd_watchdog.rs:55]: The two matches can be merged.

```suggestion
        Err(WatchdogError::NoWatchdogSec { file }) => {
```

Comment (PradeepKiruvale) [reply crates/core/tedge_watchdog/src/systemd_watchdog.rs]: fixed

**Review by didier-wenzek — APPROVED** [review]
Approved

**Pull request merged**